litho-core/src/main/java/com/facebook/litho/ComponentTree.java

```diff
@@ -1902,54 +1922,267 @@ protected LayoutState calculateLayoutState(
       @Nullable LayoutState previousLayoutState,
       @Nullable TreeProps treeProps,
       @CalculateLayoutSource int source,
-      String extraAttribution) {
-    final ComponentContext contextWithStateHandler;
-
-    synchronized (this) {
-      final KeyHandler keyHandler =
-          (ComponentsConfiguration.useGlobalKeys || ComponentsConfiguration.isDebugModeEnabled)
-              ? new KeyHandler(mContext.getLogger())
-              : null;
+      @Nullable String extraAttribution) {
 
-      contextWithStateHandler =
-          new ComponentContext(
-              context, StateHandler.acquireNewInstance(mStateHandler), keyHandler, treeProps);
+    LayoutStateFuture localLayoutStateFuture = new LayoutStateFuture(
```

Comment:
When the flag is disabled, can we not go through the new futures code at all? Does it require duplicating much?

Reply:
Done. It's not too much duplication, but it does start getting a bit awkward...